### ISSUE TITLE
How to save the trained model and use it to predict new time series data when using ROCKET?

### ISSUE THREAD

**Issue by AnthonyFang623** [timeseriesAI/tsai#404]
I use ROCKET method to create features and RidgeClassifierCV as the classifier, it worked pretty well. But I have no idea how to save the trained model and using the trained model to predict the label of a new data.

**Comment by oguiza** [comment]
Hi @AnthonyFang623,
I'd suggest you take a look [ROCKET documentation](https://timeseriesai.github.io/tsai/models.ROCKET.html).  There are some examples at the bottom. It's pretty easy. First you fit  and save it: 
```bash
X_train, y_train, X_test, y_test = get_UCR_data('NATOPS', Xdtype='float64')
cls = RocketClassifier()
cls.fit(X_train, y_train)
cls.save('RocketClassifier')
```
and the to load it: 
```bash
cls = load_rocket('RocketClassifier')
cls.score(X_test, y_test)
```

**Comment by AnthonyFang623** [comment]
Hi, @oguiza ,
I have read the documentation, but here are the differences. I use my own data instead of download online data, and create splits like this

```python
X = np.load('mydata.npy')
y = np.load('mylabel.npy')
X2d = X[:]
X3d = to3d(X2d)
splits = get_splits(y, valid_size=.2, stratify=True, random_state=23, shuffle=True)
tfms  = [None, [Categorize()]]
batch_tfms = [TSStandardize(by_sample=True)]
dls = get_ts_dls(X3d, y, splits=splits, tfms=tfms, drop_last=False, shuffle_train=False, batch_tfms=batch_tfms, bs=10_000) 
```
and create ROCKET features, then classify
```python
model = build_ts_model(ROCKET, dls=dls, n_kernels=20_000, kss=[6, 9, 12]) 
X_train, y_train = create_rocket_features(dls.train, model)
X_valid, y_valid = create_rocket_features(dls.valid, model)
from sklearn.linear_model import RidgeClassifierCV
ridge = RidgeClassifierCV(alphas=np.logspace(-9, 9, 12), normalize=True)
ridge.fit(X_train, y_train)
```
in this model, ridge.save is not callable, and when I use unlabeled data x_test only for pridiction, I couldn't create same features like X_train and X_valid

about the method you mentioned, I need to pre-split my data into x_train, x_test and y_train, y_test
```python
x_train = np.load('x_train.npy')
y_train = np.load('y_train.npy')
x_test = np.load('x_test.npy')
y_test = np.load('y_test.npy')
X_train = to3d(x_train[:])
X_test = to3d(x_test[:])
cls = RocketClassifier(num_kernels=20000, alphas=np.logspace(-9, 9, 12))
cls.fit(X_train, y_train)
cls.score(X_train,y_train), cls.score(X_test,y_test)
```
the result is not as good as in ROCKET model, I assume the reason is I didn't randomly split data like
```python
splits = get_splits(y, valid_size=.2, stratify=True, random_state=23, shuffle=True)
```
I believe if everything is right, the results should be the same. Did I miss something or something is wrong? How can I get the same result like in ROCKET model while could predict the new data.

**Comment by oguiza** [comment]
You can split the data as you want using the method I proposed. 
To simulate the creation of an array like yours I've used this: 
```bash
X, y, splits = get_UCR_data('OliveOil', split_data=False)
X = X[:, 0]
cat_tfm = CategoryMap(y, add_na=False)
y = cat_tfm.map_objs(y)
np.save('mydata.npy', X)
np.save('mylabel.npy', y)
```
In this case X is 2d (shape 60 x 570).
This is what you could use:
```bash
X = np.load('mydata.npy')
y = np.load('mylabel.npy')
splits = get_splits(y, valid_size=.2, stratify=True, random_state=23, shuffle=True)
X_train, X_valid = X3d[splits[0]].astype('float64'), X3d[splits[1]].astype('float64')
y_train, y_valid = y[splits[0]].astype('float64'), y[splits[1]].astype('float64')
cls = RocketClassifier(num_kernels=20_000, alphas=np.logspace(-9, 9, 12), normalize_input=True)
cls.fit(X_train, y_train)
cls.score(X_train,y_train), cls.score(X_valid, y_valid)
```

That method, which allows you to easily save the cls, is almost the same as: 
```bash
from sklearn.linear_model import RidgeClassifierCV
X = np.load('mydata.npy')
y = np.load('mylabel.npy')
X2d = X[:]
X3d = to3d(X2d)
splits = get_splits(y, valid_size=.2, stratify=True, random_state=23, shuffle=True)
tfms  = [None, [Categorize()]]
batch_tfms = [TSStandardize(by_sample=True)]
dls = get_ts_dls(X3d, y, splits=splits, tfms=tfms, drop_last=False, shuffle_train=False, batch_tfms=batch_tfms, bs=10_000) 
model = build_ts_model(ROCKET, dls=dls, n_kernels=20_000, kss=[6, 9, 12]) 
X_train, y_train = create_rocket_features(dls.train, model)
X_valid, y_valid = create_rocket_features(dls.valid, model)
ridge = RidgeClassifierCV(alphas=np.logspace(-9, 9, 12), normalize=True)
ridge.fit(X_train, y_train)
ridge.score(X_train,y_train), ridge.score(X_valid,y_valid)
```
The result I get is the same (100% accuracy). There's only 1 difference: RocketClassifier (that uses Rocket from sktime), doesn't allow you to pass different kernel sizes. They are fixed at [7,9,11]. I can't do anything about that as Rocket is not part of the tsai library.
That's one of the differences. The other is that kernels are random. So every time you run them, you may get a slightly different result.

**Comment by AnthonyFang623** [comment]
> You can split the data as you want using the method I proposed.

I get it, now my data works perfectly with ROCKET model, thank you!
